From 801fb8b894c75fc1e3fa0284e096ade6dcdc1110 Mon Sep 17 00:00:00 2001
From: Evan Hunt <each@isc.org>
Date: Tue, 3 Feb 2015 18:25:28 -0800
Subject: [PATCH] [master] avoid crash due to managed-key rollover

4053.	[security]	Revoking a managed trust anchor and supplying
			an untrusted replacement could cause named
			to crash with an assertion failure.
			(CVE-2015-1349) [RT #38344]

Upstream-Status: Backport

The patch comes from https://source.isc.org/git/bind9.git

(Upstream commit 801fb8b894c75fc1e3fa0284e096ade6dcdc1110)

Signed-off-by: Yue Tao <yue.tao@windriver.com>
---
 CHANGES           |    5 +++++
 lib/dns/zone.c    |    8 +++++++-
 3 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/CHANGES b/CHANGES
index d476ed1..8973477 100644
--- a/CHANGES
+++ b/CHANGES
@@ -1,3 +1,8 @@
+4053.	[security]	Revoking a managed trust anchor and supplying
+			an untrusted replacement could cause named
+			to crash with an assertion failure.
+			(CVE-2015-1349) [RT #38344]
+
 	--- 9.9.5 released ---
 
 	--- 9.9.5rc2 released ---
diff --git a/lib/dns/zone.c b/lib/dns/zone.c
index d7c7444..1bc8608 100644
--- a/lib/dns/zone.c
+++ b/lib/dns/zone.c
@@ -7648,6 +7648,12 @@ keyfetch_done(isc_task_t *task, isc_event_t *event) {
 					     namebuf, tag);
 				trustkey = ISC_TRUE;
 			}
+		} else {
+			/*
+			 * No previously known key, and the key is not
+			 * secure, so skip it.
+			 */
+			continue;
 		}
 
 		/* Delete old version */
@@ -7694,7 +7700,7 @@ keyfetch_done(isc_task_t *task, isc_event_t *event) {
 			trust_key(zone, keyname, &dnskey, mctx);
 		}
 
-		if (!deletekey)
+		if (secure && !deletekey)
 			set_refreshkeytimer(zone, &keydata, now);
 	}
 
-- 
1.7.9.5

